config/application.rb

```diff
@@ -15,10 +15,23 @@
 # require "sprockets/railtie"
 require "rails/test_unit/railtie"
 
+require "prometheus/client/data_stores/direct_file_store"
+
 # Require the gems listed in Gemfile, including any gems
 # you've limited to :test, :development, or :production.
 Bundler.require(*Rails.groups)
 
+PROMETHEUS_DIR = "/tmp/prometheus".freeze
+
+# Needs to clear before initializing.
+Dir["#{PROMETHEUS_DIR}/*.bin"].each do |file_path|
+  File.unlink(file_path)
+end
+
+# The DirectFileStore is the only way to aggregate metrics across processes.
+file_store = Prometheus::Client::DataStores::DirectFileStore.new(dir: PROMETHEUS_DIR)
```

Comment:
I'm not understanding this - but a FileStore looks like a problem? The comment references aggregating across processes, but we'll need to aggregate across nodes. The file store will also be lost on deployment?

Reply:
Aggregating across nodes should be done by the Prometheus server using jobs:

https://prometheus.io/docs/concepts/jobs_instances/

Restarting the application should only lose a small sample of metrics (from the time of the last scrape to the time the app restarts); if we scrape frequently enough the impact of this should be negligible. Granted, we're sending a potentially low-frequency metric for the feedback responses, but again the amount we miss from this edge case will be small (I'll raise it with PM to make sure they're happy with this).

The `DirectFileStore` is the recommended approach for applications with pre-fork servers/multi-processors, but there are caveats:

- When scraped by Prometheus the store reads all the files into memory (one file per metric per process), which can bump up the memory usage a fair amount. I don't anticipate this being an issue for us yet/soon given the amount of metrics we're collecting is small.

- The metric files need to be cleared out each time the application starts to avoid a new instance exporting metrics from a previous one; I've done this programatically on start but since we're dockerized it probably won't matter anyway.

- A large number of files can be created; there's no real solution for this one in the client library. If we're on the way to running out of disk space we should be alerted by infrastructure monitoring, but I imagine worst-case scenario is the container runs out of memory, blows up and a new one spins up to take it's place. It would only result in a problem if it happens on all instances at the same time - if we're worried about that then we could schedule periodic restarts of the instances to keep the usage low.